### ISSUE TITLE
صيشيشؤش

### ISSUE THREAD

**Issue by 23234wwschool-wq** [23234wwschool-wq/-#1]
# ===========================================
# كود لإنشاء نموذج ذكاء اصطناعي أساسي
# للتعرف على الأرقام المكتوبة بخط اليد (MNIST)
# ===========================================

import tensorflow as tf
from tensorflow import keras
import numpy as np
import matplotlib.pyplot as plt

print("✅ جاهز لبناء أول نموذج ذكاء اصطناعي!")
print("=" * 50)

# 1. تحميل البيانات (مجموعة MNIST الشهيرة)
print("📥 جارٍ تحميل البيانات...")
(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()

print(f"📊 شكل البيانات التدريبية: {x_train.shape}")
print(f"📊 شكل التسميات التدريبية: {y_train.shape}")

# 2. معالجة البيانات
print("\n🔄 جارٍ معالجة البيانات...")
# تحويل القيم إلى نطاق [0, 1]
x_train = x_train.astype("float32") / 255.0
x_test = x_test.astype("float32") / 255.0

# تغيير الشكل لإضافة قناة واحدة (لصور الأبيض والأسود)
x_train = np.expand_dims(x_train, -1)
x_test = np.expand_dims(x_test, -1)

# تحويل التسميات إلى ترميز one-hot
y_train = keras.utils.to_categorical(y_train, 10)
y_test = keras.utils.to_categorical(y_test, 10)

print(f"🎯 عدد الفئات: 10 (الأرقام من 0 إلى 9)")

# 3. بناء نموذج الشبكة العصبية
print("\n🏗️ جارٍ بناء نموذج الشبكة العصبية...")

model = keras.Sequential([
    keras.layers.Input(shape=(28, 28, 1)),
    
    # طبقة الإدخال والطبقات المخفية
    keras.layers.Conv2D(32, kernel_size=(3, 3), activation="relu"),
    keras.layers.MaxPooling2D(pool_size=(2, 2)),
    
    keras.layers.Conv2D(64, kernel_size=(3, 3), activation="relu"),
    keras.layers.MaxPooling2D(pool_size=(2, 2)),
    
    # تسطيح البيانات
    keras.layers.Flatten(),
    
    # طبقات كاملة الاتصال
    keras.layers.Dense(128, activation="relu"),
    keras.layers.Dropout(0.5),  # لمنع الإفراط في التطابق
    
    # طبقة الإخراج
    keras.layers.Dense(10, activation="softmax")
])

# 4. تجميع النموذج
print("🔧 جارٍ تجميع النموذج...")
model.compile(
    optimizer="adam",
    loss="categorical_crossentropy",
    metrics=["accuracy"]
)

# عرض ملخص النموذج
print("\n📋 ملخص النموذج:")
print("=" * 40)
model.summary()
print("=" * 40)

# 5. تدريب النموذج
print("\n🚀 بدء التدريب...")
print("⏳ هذا قد يستغرق دقيقتين...")

history = model.fit(
    x_train, y_train,
    batch_size=128,
    epochs=10,
    validation_split=0.1,
    verbose=1
)

# 6. تقييم النموذج
print("\n📈 تقييم النموذج على بيانات الاختبار...")
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=0)
print(f"✅ دقة النموذج على بيانات الاختبار: {test_acc:.4f} ({test_acc*100:.2f}%)")

# 7. عرض بعض النتائج
print("\n🎯 عرض بعض التنبؤات...")

# اختيار 5 عينات عشوائية
indices = np.random.choice(len(x_test), 5, replace=False)

plt.figure(figsize=(15, 3))
for i, idx in enumerate(indices):
    # الحصول على الصورة والتنبؤ
    img = x_test[idx]
    true_label = np.argmax(y_test[idx])
    
    # التنبؤ
    pred = model.predict(img[np.newaxis, ...], verbose=0)
    pred_label = np.argmax(pred)
    
    # عرض الصورة
    plt.subplot(1, 5, i+1)
    plt.imshow(img.squeeze(), cmap="gray")
    plt.title(f"الحقيقي: {true_label}\nالتنبؤ: {pred_label}")
    plt.axis("off")

plt.tight_layout()
plt.show()

# 8. حفظ النموذج
print("\n💾 جارٍ حفظ النموذج...")
model.save("my_first_ai_model.h5")
print("✅ تم حفظ النموذج في: 'my_first_ai_model.h5'")

# 9. دالة للتنبؤ بصورة جديدة
def predict_digit(image):
    """
    دالة للتنبؤ برقم في صورة
    """
    if image.shape != (28, 28, 1):
        # تغيير حجم الصورة إذا لزم الأمر
        image = tf.image.resize(image, (28, 28))
        image = np.expand_dims(image, -1)
    
    # التنبؤ
    prediction = model.predict(image[np.newaxis, ...], verbose=0)
    digit = np.argmax(prediction)
    confidence = np.max(prediction)
    
    return digit, confidence

print("\n" + "=" * 50)
print("🎉 مبروك! لقد بنيت أول نموذج ذكاء اصطناعي!")
print("=" * 50)
print("\n📌 ملاحظات:")
print("1. النموذج يتعرف على الأرقام المكتوبة بخط اليد")
print("2. يمكنك تحسينه بإضافة المزيد من البيانات أو الطبقات")
print("3. جرب تغيير معلمات النموذج لتحسين الدقة")
print("4. النموذج محفوظ في 'my_first_ai_model.h5'")

# 10. مثال على استخدام النموذج
print("\n🔍 مثال على استخدام النموذج:")
sample_image = x_test[0]
digit, confidence = predict_digit(sample_image)
print(f"   الصورة الأولى في مجموعة الاختبار هي الرقم: {digit}")
print(f"   درجة الثقة: {confidence:.2%}")
